docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,841 @@
+
+Each module is represented by a `ModuleNode` instance. Modules may be registered in the graph without yet being processed (`transformResult` would be `null` in that case). `importers` and `importedModules` are also updated after the module is processed.
+
+```ts
+class ModuleNode {
+  environment: string
+
+  url: string
+  id: string | null = null
+  file: string | null = null
+
+  type: 'js' | 'css'
+
+  importers = new Set<ModuleNode>()
+  importedModules = new Set<ModuleNode>()
+  importedBindings: Map<string, Set<string>> | null = null
+
+  info?: ModuleInfo
+  meta?: Record<string, any>
+  transformResult: TransformResult | null = null
+
+  acceptedHmrDeps = new Set<ModuleNode>()
+  acceptedHmrExports: Set<string> | null = null
+  isSelfAccepting?: boolean
+  lastHMRTimestamp = 0
+  lastInvalidationTimestamp = 0
+}
+```
+
+`environment.moduleGraph` is an instance of `ModuleGraph`:
+
+```ts
+export class ModuleGraph {
+  environment: string
+
+  urlToModuleMap = new Map<string, ModuleNode>()
+  idToModuleMap = new Map<string, ModuleNode>()
+  etagToModuleMap = new Map<string, ModuleNode>()
+  fileToModulesMap = new Map<string, Set<ModuleNode>>()
+
+  constructor(
+    environment: string,
+    resolveId: (url: string) => Promise<PartialResolvedId | null>,
+  )
+
+  async getModuleByUrl(rawUrl: string): Promise<ModuleNode | undefined>
+
+  getModulesByFile(file: string): Set<ModuleNode> | undefined
+
+  onFileChange(file: string): void
+
+  invalidateModule(
+    mod: ModuleNode,
+    seen: Set<ModuleNode> = new Set(),
+    timestamp: number = Date.now(),
+    isHmr: boolean = false,
+  ): void
+
+  invalidateAll(): void
+
+  async updateModuleInfo(
+    mod: ModuleNode,
+    importedModules: Set<string | ModuleNode>,
+    importedBindings: Map<string, Set<string>> | null,
+    acceptedModules: Set<string | ModuleNode>,
+    acceptedExports: Set<string> | null,
+    isSelfAccepting: boolean,
+  ): Promise<Set<ModuleNode> | undefined>
+
+  async ensureEntryFromUrl(
+    rawUrl: string,
+    setIsSelfAccepting = true,
+  ): Promise<ModuleNode>
+
+  createFileOnlyEntry(file: string): ModuleNode
+
+  async resolveUrl(url: string): Promise<ResolvedUrl>
+
+  updateModuleTransformResult(
+    mod: ModuleNode,
+    result: TransformResult | null,
+  ): void
+
+  getModuleByEtag(etag: string): ModuleNode | undefined
+}
+```
+
+## Creating new environments
+
+One of the goals of this feature is to provide a customizable API to process and run code. Users can create new environment types using the exposed primitives.
+
+```ts
+import { DevEnvironment } from 'vite'
+
+function createWorkerdDevEnvironment(server: ViteDevServer, name: string, config?: DevEnvironmentConfig) {
+  const workerdModuleRunner = /* ... */
```

Comment:
I think a runner should not be declared here. What it should be here is a thing that connects the node process and the workerd process.

Or does "runner" and "moduleRunner" have a different meaning? If then, I think we should have a explanation of that.

- "runner": Always on the node process running vite. It uses a module runner in the same process (Local Runner) or a different process to run modules (Remote Runner).
- "module runner": Might be on a different process. It connects the modules and uses the ModuleEvaluator to evaluate them.

Reply:
This can just be removed from here. The initial idea was to introduce a proxy, but we decided against it after talking with cloudflare folks because it doesn't address the main issue of calling a server on another machine (so, the ssr primitive was wrong - it's not a module but a middleware).

I will replace it with the new `RemoteEnvironmentTransport` example